File: packages/commerce-types/tests/multi-object.test.ts
import { describe, expect, it } from "vitest";

import { createSession } from "../src/session.js";
import { commitmentVersion, type World } from "../src/guard.js";
import { checkI6TreeConsistency } from "../src/invariants.js";
import { commitmentId, newCommitment, partyId, valueId, type Commitment, type Value } from "../src/primitives.js";
import { applyCommitmentPath } from "../src/transitions.js";

const buyer = partyId("buyer");
const seller = partyId("seller");

function moneyValue(amount: number): Value {
  return { id: valueId(), form: { kind: "Money", money: { amount, currency: "MAD" } }, quantity: 1, state: { type: "Available" } };
}

/** A Fulfilled commitment with id/amount and optional parent/children links. */
function commit(id: string, amount: number, links: Partial<Pick<Commitment, "parent" | "children">> = {}): Commitment {
  const base = { ...newCommitment(buyer, seller, { offered: [], requested: [moneyValue(amount)] }), id: commitmentId(id), ...links };
  return applyCommitmentPath(base, { type: "Fulfilled" }, seller);
}

/** A parent (200) with two 100-children that reconcile via I-6. */
function tree(parentId = "order-1"): Commitment[] {
  const a = `${parentId}-A`;
  const b = `${parentId}-B`;
  return [
    commit(parentId, 200, { children: [commitmentId(a), commitmentId(b)] }),
    commit(a, 100, { parent: commitmentId(parentId) }),
    commit(b, 100, { parent: commitmentId(parentId) }),
  ];
}

function refund(commitment: string, amount: number, key: string) {
  return { commitment, to: { type: "Refunded" as const, amount: { amount, currency: "MAD" }, at: "2026-02-01T00:00:00.000Z" }, actor: seller, idempotencyKey: key };
}

describe("createSession — multi-object coherence (per-tree cumulative)", () => {
  it("catches a cumulative over-refund spread across the tree (the gap)", () => {
    const [parent, a, b] = tree();
    expect(checkI6TreeConsistency(parent!, [a!, b!])).toHaveLength(0); // I-6 reconciles
    const session = createSession({ commitments: [parent!, a!, b!], fulfillments: [], parties: [] });

    expect(session.propose(refund("order-1-A", 80, "a")).ok).toBe(true); // child ≤ 100
    expect(session.propose(refund("order-1-B", 80, "b")).ok).toBe(true); // child ≤ 100
    const over = session.propose(refund("order-1", 80, "p")); // parent ≤ 200, but tree → 240
    expect(over.ok).toBe(false);
    if (!over.ok) {
      expect(over.violations[0]?.rule).toBe("I-1");
      expect(over.violations[0]?.message).toContain("tree");
      expect(over.violations[0]?.message).toContain("240");
      expect(over.violations[0]?.message).toContain("200");
      const alt = (over.alternatives ?? []).find((x) => x.to === "Refunded");
      expect(alt?.bounded).toContain("40"); // remaining across the tree
    }
  });

  it("each child refund is individually valid (≤ its own committed) yet the tree caps the sum", () => {
    const [parent, a, b] = tree();
    const session = createSession({ commitments: [parent!, a!, b!], fulfillments: [], parties: [] });
    // child refunds within their own committed (100) are fine…
    expect(session.propose(refund("order-1-A", 100, "a")).ok).toBe(true);
    expect(session.propose(refund("order-1-B", 100, "b")).ok).toBe(true);
    // …but the tree is now fully refunded; any further refund anywhere in it is capped.
    const more = session.propose(refund("order-1", 1, "p"));
    expect(more.ok).toBe(false);
  });

  it("a child over-refund is still caught per-commitment (per-child cap intact)", () => {
    const [parent, a, b] = tree();
    const session = createSession({ commitments: [parent!, a!, b!], fulfillments: [], parties: [] });
    // child-A committed 100; refunding 150 against it exceeds the CHILD itself.
    const over = session.propose(refund("order-1-A", 150, "a"));
    expect(over.ok).toBe(false);
    if (!over.ok) expect(over.violations[0]?.rule).toBe("I-1");
  });

  it("single-commitment (no tree) behaviour is unchanged", () => {
    const standalone = commit("solo", 200); // no parent, no children
    const session = createSession({ commitments: [standalone], fulfillments: [], parties: [] });
    expect(session.propose(refund("solo", 120, "a")).ok).toBe(true);
    const over = session.propose(refund("solo", 100, "b")); // 220 > 200 — per-commitment
    expect(over.ok).toBe(false);
    if (!over.ok) {
      expect(over.violations[0]?.rule).toBe("I-1");
      // standalone message is the per-commitment form (not the tree form).
      expect(over.violations[0]?.message).not.toContain("tree");
    }
  });

  it("a valid tree of refunds within the parent completes", () => {
    const [parent, a, b] = tree();
    const session = createSession({ commitments: [parent!, a!, b!], fulfillments: [], parties: [] });
    expect(session.propose(refund("order-1-A", 100, "a")).ok).toBe(true);
    expect(session.propose(refund("order-1-B", 100, "b")).ok).toBe(true);
    // both children fully refunded; the tree total (200) equals the parent committed.
  });

  it("F4 replay across the tree dedups (tree ledger not double-counted)", () => {
    const [parent, a, b] = tree();
    const session = createSession({ commitments: [parent!, a!, b!], fulfillments: [], parties: [] });
    expect(session.propose(refund("order-1-A", 80, "k")).ok).toBe(true);
    const replay = session.propose(refund("order-1-A", 80, "k")); // same key
    expect(replay.ok).toBe(true);
    if (replay.ok) expect(replay.replay).toBe(true);
    // tree not double-counted: a further 120 still fits (80 + 120 = 200).
    expect(session.propose(refund("order-1-B", 120, "b")).ok).toBe(false); // child-B committed only 100
    expect(session.propose(refund("order-1-B", 100, "b2")).ok).toBe(true); // 80 + 100 = 180 ≤ 200
  });

  it("F3 conflict on a child still works (stale version)", () => {
    const acceptedChild = applyCommitmentPath({ ...newCommitment(buyer, seller, { offered: [], requested: [moneyValue(100)] }), id: commitmentId("c-1"), parent: commitmentId("p-1") }, { type: "Accepted" }, seller);
    const parent = commit("p-1", 200, { children: [commitmentId("c-1"), commitmentId("c-2")] });
    const sibling = commit("c-2", 100, { parent: commitmentId("p-1") });
    const session = createSession({ commitments: [parent, acceptedChild, sibling], fulfillments: [], parties: [] });
    const planned = commitmentVersion(session.world.commitments.find((c) => (c.id as string) === "c-1")!);
    // advance the child…
    expect(session.propose({ commitment: "c-1", to: { type: "Active" }, actor: seller, expectedVersion: planned, idempotencyKey: "x" }).ok).toBe(true);
    // …a stale-version action on the same child conflicts.
    const stale = session.propose({ commitment: "c-1", to: { type: "Disputed", by: buyer, reason: "x", opened_at: "2026-03-01T00:00:00.000Z" }, actor: buyer, expectedVersion: planned, idempotencyKey: "y" });
    expect(stale.ok).toBe(false);
    if (!stale.ok) expect(stale.conflict).toBe(true);
  });
});

File: bindings/go/multi_object_test.go
package warp

import (
	"encoding/json"
	"fmt"
	"strings"
	"testing"

	gen "github.com/yasirlts/warp-lang/bindings/go/generated"
)

// commitLinked builds a Fulfilled commitment with optional parent / children links.
func commitLinked(id string, amount float64, parent string, children []string) gen.Commitment {
	parentJSON := "null"
	if parent != "" {
		parentJSON = fmt.Sprintf("%q", parent)
	}
	childrenJSON, _ := json.Marshal(children)
	if children == nil {
		childrenJSON = []byte("[]")
	}
	var c gen.Commitment
	j := fmt.Sprintf(`{"id":"%s","parties":{"initiator":"buyer","counterparty":"seller","intermediaries":[]},"subject":{"offered":[],"requested":[{"id":"v","form":{"kind":"Money","money":{"amount":%v,"currency":"MAD"}},"quantity":1,"state":{"type":"Available"}}]},"state":{"type":"Fulfilled"},"history":[],"parent":%s,"children":%s,"created_at":"2026-01-02T08:00:00.000Z"}`, id, amount, parentJSON, string(childrenJSON))
	if err := json.Unmarshal([]byte(j), &c); err != nil {
		panic(err)
	}
	return c
}

// tree builds a parent (200) with two 100-children that reconcile via I-6.
func tree(parentID string) []gen.Commitment {
	a := parentID + "-A"
	b := parentID + "-B"
	return []gen.Commitment{
		commitLinked(parentID, 200, "", []string{a, b}),
		commitLinked(a, 100, parentID, nil),
		commitLinked(b, 100, parentID, nil),
	}
}

func TestTreeCatchesCumulativeOverRefundSpreadAcrossChildren(t *testing.T) {
	s := CreateSession(World{Commitments: tree("order-1")})
	// Each child refund is <= its own committed (100) — individually valid.
	if !s.Propose(refundActionKeyed("order-1-A", 80, "a")).OK {
		t.Fatal("child-A refund 80 should be accepted")
	}
	if !s.Propose(refundActionKeyed("order-1-B", 80, "b")).OK {
		t.Fatal("child-B refund 80 should be accepted")
	}
	// Parent refund of 80 <= 200 on its own, but the TREE would reach 240 > 200.
	over := s.Propose(refundActionKeyed("order-1", 80, "p"))
	if over.OK {
		t.Fatal("expected the tree cap to reject the third refund")
	}
	if over.Violations[0].Rule != "I-1" {
		t.Fatalf("expected I-1, got %s", over.Violations[0].Rule)
	}
	if !strings.Contains(over.Violations[0].Message, "tree") {
		t.Fatalf("message %q should mention the tree", over.Violations[0].Message)
	}
	if !strings.Contains(over.Violations[0].Message, "240") || !strings.Contains(over.Violations[0].Message, "200") {
		t.Fatalf("message %q should mention 240 and 200", over.Violations[0].Message)
	}
	if over.Alternatives[0].Bounded == nil || !strings.Contains(*over.Alternatives[0].Bounded, "40") {
		t.Fatal("bounded alternative should report 40 remaining across the tree")
	}
}

func TestTreeCapsSumEvenWhenEachChildIsIndividuallyValid(t *testing.T) {
	s := CreateSession(World{Commitments: tree("order-1")})
	if !s.Propose(refundActionKeyed("order-1-A", 100, "a")).OK {
		t.Fatal("child-A refund 100 should be accepted")
	}
	if !s.Propose(refundActionKeyed("order-1-B", 100, "b")).OK {
		t.Fatal("child-B refund 100 should be accepted")
	}
	// The tree is now fully refunded; any further refund in it is capped.
	more := s.Propose(refundActionKeyed("order-1", 1, "p"))
	if more.OK {
		t.Fatal("a further refund on a fully-refunded tree should be rejected")
	}
}

func TestChildOverRefundStillCaughtPerCommitment(t *testing.T) {
	s := CreateSession(World{Commitments: tree("order-1")})
	// child-A committed 100; refunding 150 against it exceeds the CHILD itself.
	over := s.Propose(refundActionKeyed("order-1-A", 150, "a"))
	if over.OK {
		t.Fatal("per-child over-refund should be rejected")
	}
	if over.Violations[0].Rule != "I-1" {
		t.Fatalf("expected I-1, got %s", over.Violations[0].Rule)
	}
}

func TestStandaloneCommitmentUnchangedMessageIsPerCommitmentForm(t *testing.T) {
	s := CreateSession(World{Commitments: []gen.Commitment{fulfilledOrder("solo", 200)}})
	if !s.Propose(refundActionKeyed("solo", 120, "a")).OK {
		t.Fatal("standalone refund 120 should be accepted")
	}
	over := s.Propose(refundActionKeyed("solo", 100, "b")) // 220 > 200
	if over.OK {
		t.Fatal("standalone cumulative over-refund should be rejected")
	}
	if over.Violations[0].Rule != "I-1" {
		t.Fatalf("expected I-1, got %s", over.Violations[0].Rule)
	}
	// The standalone message is the per-commitment form, not the tree form.
	if strings.Contains(over.Violations[0].Message, "tree") {
		t.Fatalf("standalone message %q should not mention a tree", over.Violations[0].Message)
	}
}

func TestValidTreeOfRefundsWithinTheParentCompletes(t *testing.T) {
	s := CreateSession(World{Commitments: tree("order-1")})
	if !s.Propose(refundActionKeyed("order-1-A", 100, "a")).OK {
		t.Fatal("child-A refund 100 should be accepted")
	}
	if !s.Propose(refundActionKeyed("order-1-B", 100, "b")).OK {
		t.Fatal("child-B refund 100 should be accepted")
	}
}

func TestTreeReplayDoesNotDoubleCountTheTreeLedger(t *testing.T) {
	s := CreateSession(World{Commitments: tree("order-1")})
	if !s.Propose(refundActionKeyed("order-1-A", 80, "k")).OK {
		t.Fatal("child-A refund 80 should be accepted")
	}
	replay := s.Propose(refundActionKeyed("order-1-A", 80, "k")) // same key
	if !replay.Replay {
		t.Fatal("same-key retry should be a replay")
	}
	// Tree not double-counted: child-B committed only 100, so 120 fails per-child…
	if s.Propose(refundActionKeyed("order-1-B", 120, "b")).OK {
		t.Fatal("child-B 120 should fail per-child cap")
	}
	// …but 100 fits (tree 80 + 100 = 180 <= 200).
	if !s.Propose(refundActionKeyed("order-1-B", 100, "b2")).OK {
		t.Fatal("child-B 100 should fit within the tree (80 + 100 = 180 <= 200)")
	}
}
